### PR TITLE
Add float singleton type

### LINKED ISSUE #284: Add float singleton types by generifying string type logic
Generify string type union, intersect, diff algorithms by passing a comparator. 
Use that to create float singleton type.

Part of #262.

### PR DESCRIPTION
**Pull request by manuranga**
Fixes #284

### CODE REVIEW

File: compiler/modules/types/flaot.bal [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+       return createFloatSubtype(false, floatListDiff(t2.values, t1.values));
+    }
+    else {
+        // !b1 && b2
+       return createFloatSubtype(false, floatListDiff(t1.values, t2.values));
+    }
+}
+
+function floatSubtypeIntersect(SubtypeData d1, SubtypeData d2) returns SubtypeData {
+    FloatSubtype t1 = <FloatSubtype>d1;
```
Comment (manuranga) [inline compiler/modules/types/flaot.bal]: Should I generify this part by creating `type ListableSubtype FloatSubtype|StringSubtype`

File: compiler/modules/types/listable.bal [hunk 2]
```diff
@@ -0,0 +1,148 @@
+// ListableTypes are types in which each expressible subtype can be reasonably
+// represented by a list of values. counter-eg: `int` is not a ListableType since
+// uint32 can't be reasonably represented by listing all values.
+type ListableType float|string;
```
Comment (manuranga) [inline compiler/modules/types/listable.bal]: is `ListableType` a good name for this ?

Comment (jclark) [reply compiler/modules/types/listable.bal]: I suggest `EnumerableType`

File: compiler/modules/types/flaot.bal (hunk 1 above)
Comment (jclark) [reply compiler/modules/types/flaot.bal]: Yes, `EnumerableSubtype`

File: compiler/modules/types/flaot.bal [hunk 3]
```diff
@@ -0,0 +1,98 @@
+// Implementation specific to basic type float.
```
Comment (jclark) [inline compiler/modules/types/flaot.bal]: Filename is misspelled

File: compiler/modules/types/listable.bal [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+const LT = -1;
+const EQ = 0;
+const GT = 1;
+
+type Order GT|EQ|LT;
+
+function listUnion(ListableType[] v1, ListableType[] v2, ListableType[] result) {
```
Comment (manuranga) [inline compiler/modules/types/listable.bal]: Accepting `result` as a pram feels bit unclean, but this avoids a cast.

Comment (jclark) [reply compiler/modules/types/listable.bal]: I think adding the param is the right choice, but I would add a word (maybe `into`) to the name to make it clear that this is not working functionally (unlike `int`).

**Comment by manuranga** [comment]
Ready to be reviewed.
Tests don't cover  list intersect, I think we need `!` type constructor to do that easily.

File: compiler/modules/types/listable.bal (hunk 4 above)
Comment (manuranga) [reply compiler/modules/types/listable.bal]: I prefixed `add`, could have gone with `collect` or `accumulate`.

File: compiler/modules/types/float.bal [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    boolean allowed;
+    float[] values;
+|};
+
+public function floatConst(float value) returns SemType {
+    FloatSubtype st = { allowed: true, values: [value] };
+    return uniformSubtype(UT_FLOAT, st);
+}
+
+function floatSubtypeUnion(SubtypeData d1, SubtypeData d2) returns SubtypeData =>
```
Comment (jclark) [inline compiler/modules/types/float.bal]: Don't see why we need this. Can't we just put enumerableSubtypeUnion in the vtable?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/types/enumerable.bal [hunk 6]
```diff
@@ -0,0 +1,225 @@
+import wso2/nballerina.err;
+
+// EnumerableTypes are types in which each expressible subtype can be reasonably
+// represented by a list of values. counter-eg: `int` is not a EnumerableType since
+// uint32 can't be reasonably represented by listing all values.
+type EnumerableType float|string;
+
+type EnumerableSubType FloatSubtype|StringSubtype;
```
Comment (jclark) [inline compiler/modules/types/enumerable.bal:8]: It's `Subtype` not `SubType`.

File: compiler/modules/types/string.bal [hunk 7]
```diff
@@ -194,10 +50,12 @@ function createStringSubtype(boolean allowed, string[] values) returns SubtypeDa
 }
 
 final UniformTypeOps stringOps = {
-    union: stringSubtypeUnion,
+    union: function (SubtypeData d1, SubtypeData d2) returns SubtypeData =>
```
Comment (jclark) [inline compiler/modules/types/string.bal:53]: Just put enumerableSubtypeUnion directly in the ops table (and make the parameter type be SubtypeData). Similarly for all the other functions in the ops table.

**Review by jclark — CHANGES_REQUESTED** [review]

**Comment by jclark** [comment]
Also finish this off by

- making resolveTypes use t:floatConst
- adding a float-singleton.bal showTypes test case

**Comment by jclark** [comment]
We can tweak this so that it will work with string:Char (which will need additional members in StringSubtype) as follows:

```
function enumerableSubtypeUnion(EnumerableSubType t1, EnumerableSubType t2) returns SubtypeData {}
```

becomes

```
function enumerableSubtypeUnion(EnumerableSubtype t1, EnumerableSubtype t2, EnumerableSubtype result) {}
```

where EnumerableSubtype is an open record:

```
type EnumerableSubtype readonly & record {
    boolean allowed;
    EnumerableType[] values;
};
```

We will then still need stringSubtypeUnion and floatSubtypeUnion as trivial wrappers around the corresponding enumerable* function.

Please don't use anonymous functions in the ops table.

**Comment by manuranga** [comment]
In order to write showTypes for NaN and Infinity, I'll need #293.

(eg: 
```ballerina
const float NAN = -1.0/0.0;
type NaN NAN
```
)

**Comment by jclark** [comment]
You don't need to be able to put type descriptors in const declarations. You can just say

```
const NaN = 1.0/0.0;
```

This requires expressions to allow floating point literals. Added as #294.

**Comment by manuranga** [comment]
> You don't need to be able to put type descriptors in const declarations.

Great, I was misled since jBallerina required that. Guess that's a spec deviation.

**Comment by jclark** [comment]
> Tests don't cover list intersect, I think we need `!` type constructor to do that easily.

Yeah, we definitely need `!`.

**Pull request merged**